src/meta-srv/src/selector/load_based.rs

```diff
@@ -12,79 +12,223 @@
+        check(&stat_kvs, &opts)?;
+
+        // 5. compute weight array
+        let weight_array = compute_weight_array(&stat_kvs, &lease_kvs);
+
+        // 6. choose peers from weight_array
+        let mut weighted_choose = self.weighted_choose.write();
+        let selected = choose_peers(weight_array, &opts, &mut *weighted_choose);
+
+        Ok(selected)
+    }
+}
+
+pub(crate) fn filter_out_expired_datanode(
+    mut stat_kvs: HashMap<StatKey, StatValue>,
+    lease_kvs: &HashMap<LeaseKey, LeaseValue>,
+) -> HashMap<StatKey, StatValue> {
+    lease_kvs
+        .iter()
+        .filter_map(|(lease_k, _)| stat_kvs.remove_entry(&lease_k.into()))
+        .collect()
+}
+
+pub(crate) fn filter_out_datanode_by_table(
+    stat_kvs: HashMap<StatKey, StatValue>,
+    catalog: &str,
+    schema: &str,
+    table: &str,
+) -> HashMap<StatKey, StatValue> {
+    stat_kvs
+        .into_iter()
+        .filter(|(_, stat_v)| contains_table(stat_v, catalog, schema, table) != Some(true))
+        .collect()
+}
+
+pub(crate) fn check(stat_kvs: &HashMap<StatKey, StatValue>, opts: &SelectorOptions) -> Result<()> {
+    if opts.allow_duplication {
+        ensure!(
+            !stat_kvs.is_empty(),
+            error::NoEnoughAvailableDatanodeSnafu {
+                expected: opts.min_item_num,
+                available: 0_usize
+            }
+        );
+    } else {
+        ensure!(
+            opts.min_item_num <= stat_kvs.len(),
+            error::NoEnoughAvailableDatanodeSnafu {
+                expected: opts.min_item_num,
+                available: stat_kvs.len()
+            }
+        );
+    }
+
+    Ok(())
+}
+
+/// Compute a weight for each datanode according to stat_kvs.
+///
+/// The method of compute the weight is to reverse the region number array. An example below.
+///
+/// region number array:
+/// datanode_1: 10,  datanode_2: 12, datanode_3: 20
+///
+/// weight array:
+/// datanode_1: 20,  datanode_2: 12, datanode_3: 10
+pub(crate) fn compute_weight_array(
+    stat_kvs: &HashMap<StatKey, StatValue>,
+    lease_kvs: &HashMap<LeaseKey, LeaseValue>,
+) -> Vec<WeightedItem<Peer>> {
+    let mut region_nums = Vec::with_capacity(stat_kvs.len());
+    let mut peers = Vec::with_capacity(stat_kvs.len());
+
+    for stat_k in stat_kvs.keys() {
+        let region_num = match stat_kvs
+            .get(stat_k)
+            .and_then(|stat_val| stat_val.region_num())
+        {
+            Some(region_num) => region_num,
+            None => {
+                warn!("Failed to get stat_val by stat_key {:?}", stat_k);
+                MAX_REGION_NUMBER
+            }
+        };
+
+        let Some(lease_v) = lease_kvs.get(&stat_k.into()) else {
+            continue;
+        };
+
+        let peer = Peer {
+            id: stat_k.node_id,
+            addr: lease_v.node_addr.clone(),
+        };
+
+        region_nums.push(region_num);
+        peers.push(peer);
+    }
+    region_nums.reverse();
```

Comment:
```
region_nums.reverse();
let weights = region_nums;`
```
I think it's easier to understand this way, it's like it's saying:  The number of regions is inversely proportional to the weight.